x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,162 @@
+        int currentMax = MIN_RETRY_SLEEP_MILLIS;
+        int currentAttempt = 0;
+        while(currentAttempt <= maxFailureRetries) {
+            BulkResponse bulkResponse = bulkIndex(bulkRequest);
+            if (bulkResponse.hasFailures() == false) {
+                return bulkResponse;
+            }
+            if (shouldRetry.get() == false) {
+                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+            }
+            if (currentAttempt > maxFailureRetries) {
+                LOGGER.warn("[{}] failed to index after [{}] attempts.", jobId, currentAttempt);
+                throw new ElasticsearchException("[{}] failed to index all results after [{}] attempts. {}",
+                    jobId,
+                    currentAttempt,
+                    bulkResponse.buildFailureMessage());
+            }
+            currentAttempt++;
+            // Since we exponentially increase, we don't want force randomness to have an excessively long sleep
+            if (currentMax < MAX_RETRY_SLEEP_MILLIS) {
+                currentMin = currentMax;
+            }
+            double backOff = ((1 << Math.min(currentAttempt, MAX_RETRY_EXPONENT)) - 1) / 2.0;
+            int max = (int)(backOff * 100);
+            currentMax = Math.min(max, MAX_RETRY_SLEEP_MILLIS);
+            // Its good to have a random window along the exponentially increasing curve
+            // so that not all bulk requests rest for the same amount of time
+            int randSleep = currentMin + random.nextInt(1 + (currentMax - currentMin));
+            // We should only retry the docs that failed.
+            bulkRequest = buildNewRequestFromFailures(bulkRequest, bulkResponse);
+            try {
+                Thread.sleep(randSleep);
+            } catch (InterruptedException interruptedException) {
+                LOGGER.warn(
+                    new ParameterizedMessage("[{}] failed to index after [{}] attempts due to interruption",
+                        jobId,
+                        currentAttempt),
+                    interruptedException);
+                Thread.currentThread().interrupt();
+            }
+        }
+        throw new ElasticsearchException("[{}] failed to index all results after [{}] attempts. {}",
+            jobId,
+            currentAttempt);
+    }
+
+    private BulkResponse bulkIndex(BulkRequest bulkRequest) {
+        try (ThreadContext.StoredContext ignore = client.threadPool().getThreadContext().stashWithOrigin(ML_ORIGIN)) {
+            return client.bulk(bulkRequest).actionGet();
+        }
+    }
+
+    private BulkRequest buildNewRequestFromFailures(BulkRequest bulkRequest, BulkResponse bulkResponse) {
+        // If we failed, lets set the bulkRequest to be a collection of the failed requests
+        BulkRequest bulkRequestOfFailures = new BulkRequest();
+        Set<String> failedDocIds = new HashSet<>();
```

Comment:
Could be expressed as a one declarative statement, something like:
```
Set<String> failedDocIds = bulkResponse.stream().filter(::isFailed).map(::getId).collect(toSet());
```

This way, you can avoid mutable code (i.e. calling `HashSet::add`) entirely.